thumbfast.lua

```diff
@@ -888,7 +1018,318 @@ local function sync_changes(prop, val)
+        end
+    )
+
+    fetch_fragment(storyboard, i+1, thumbnail_size, storyboard_scale)
+end
+
+local function anycase(s)
+    return string.gsub(s, "%a", function (c)
+        return string.format("[%s%s]", c:lower(), c:upper())
+    end)
+end
+
+local http_prefix = anycase("^https?://")
+local ytdl_prefix = "^ytdl://(.+)"
+local subdomains = "[%w-.]*"
+local naked_ytdl_id = "^ytdl://([%w-_]+)$"
+local youtube_id = "[%w-_]+.*"
+local twitch_id = "%d+.*"
+local ytdl_opts = {try_ytdl_first = false, ytdl_path = ""}
+mp.options.read_options(ytdl_opts, "ytdl_hook")
+
+local youtube_patterns_free = {
+    -- youtube.com/watch?v=abcdef01234 or any invidious/piped site
+    anycase(".*/watch.*[?&]v=")..youtube_id,
+
+    -- youtube.com/embed/abcdef01234 or any invidious/piped site
+    anycase(".*/embed/")..youtube_id,
+}
+local youtube_patterns = {
+    -- youtu.be/abcdef01234
+    "^"..anycase("youtu%.be/")..youtube_id,
+
+    -- youtube.com/v/abcdef01234
+    "^"..subdomains..anycase("youtube%.com/[^/]+/")..youtube_id,
+}
+local twitch_base = subdomains..anycase("twitch%.tv/")
+local twitch_patterns = {
+    -- twitch.tv/user/v/123456
+    "^"..subdomains..anycase("twitch%.tv/[^/]+/v/")..twitch_id,
+
+    -- twitch.tv/user/video/123456
+    "^"..subdomains..anycase("twitch%.tv/[^/]+/video/")..twitch_id,
+
+    -- twitch.tv/videos/123456
+    "^"..subdomains..anycase("twitch%.tv/videos/")..twitch_id,
+
+    -- twitch.tv/user/schedule?vodID=123456
+    "^"..subdomains..anycase("twitch%.tv/[^/]+/schedule%?vodID=")..twitch_id,
+
+    -- player.twitch.tv/?video=v123456 or player.twitch.tv/?video=123456
+    "^"..anycase("player%.twitch%.tv/.*[?&]video=v?")..twitch_id,
+}
+
+local function storyboard_supported_url(path, referer)
+    local video_url = string.match(path, naked_ytdl_id) or string.match(referer, naked_ytdl_id)
+    if video_url then
+        return video_url
+    end
+
+    path_ytdl, path_has_ytdl_prefix = string.gsub(path, ytdl_prefix, "%1")
+    path,      path_has_http_prefix = string.gsub(path_ytdl, http_prefix, "")
+    path_has_prefix = path_has_ytdl_prefix or path_has_http_prefix or ytdl_opts.try_ytdl_first
+
+    referer_ytdl, referer_has_ytdl_prefix = string.gsub(referer, ytdl_prefix, "%1")
+    referer,      referer_has_http_prefix = string.gsub(referer_ytdl, http_prefix, "")
+    referer_has_prefix = referer ~= "" and (referer_has_ytdl_prefix or referer_has_http_prefix or ytdl_opts.try_ytdl_first)
+
+    local checks = {
+        {input = path_ytdl,    patterns = youtube_patterns_free, condition = function(input) return true end},
+        {input = path,         patterns = youtube_patterns,      condition = function(input) return path_has_prefix end},
+        {input = path,         patterns = twitch_patterns,       condition = function(input) return path_has_prefix and string.match(input, twitch_base) end},
+        {input = referer_ytdl, patterns = youtube_patterns_free, condition = function(input) return input ~= "" end},
```

Comment:
These `*_ytdl` variables just seem to be longer versions of the stripped ones, so why check against them?
```suggestion
        {input = path,    patterns = youtube_patterns_free, condition = function(input) return true end},
        {input = path,         patterns = youtube_patterns,      condition = function(input) return path_has_prefix end},
        {input = path,         patterns = twitch_patterns,       condition = function(input) return path_has_prefix and string.match(input, twitch_base) end},
        {input = referer, patterns = youtube_patterns_free, condition = function(input) return input ~= "" end},
```

Reply:
The freeform patterns in `youtube_patterns_free` match strings with anything at the start. Because the user may have prefixed the path with `ytdl://` to force ytdl on their end, which is an mpv invention that the yt-dlp command line doesn't understand, we have to make sure we're matching on the version on the path that we've stripped the ytdl:// prefix from.